src/libostree/ostree-sign-ed25519.c

```diff
@@ -0,0 +1,588 @@
+    }
+
+  *out_secret_key = g_variant_new_fixed_array (G_VARIANT_TYPE_BYTE, sk, crypto_sign_SECRETKEYBYTES, sizeof(guchar));
+  *out_public_key = g_variant_new_fixed_array (G_VARIANT_TYPE_BYTE, pk, crypto_sign_PUBLICKEYBYTES, sizeof(guchar));
+
+  return TRUE;
+#endif /* HAVE_LIBSODIUM */
+
+err:
+  return FALSE;
+}
+
+gboolean ostree_sign_ed25519_set_sk (OstreeSign *self,
+                                     GVariant *secret_key,
+                                     GError **error)
+{
+  g_debug ("%s enter", __FUNCTION__);
+  g_return_val_if_fail (OSTREE_IS_SIGN (self), FALSE);
+
+#ifdef HAVE_LIBSODIUM
+  OstreeSignEd25519 *sign = ostree_sign_ed25519_get_instance_private(OSTREE_SIGN_ED25519(self));
+  g_autofree char * hex = NULL;
+
+  g_free (sign->secret_key);
+
+  gsize n_elements = 0;
+  sign->secret_key = (guchar *) g_variant_get_fixed_array (secret_key, &n_elements, sizeof(guchar));
+
+  if (n_elements != crypto_sign_SECRETKEYBYTES)
+    {
+      g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                           "Incorrect ed25519 secret key");
+      goto err;
+    }
+
+  hex = g_malloc0 (crypto_sign_SECRETKEYBYTES*2 + 1);
+//  g_debug ("Set ed25519 secret key = %s", sodium_bin2hex (hex, crypto_sign_SECRETKEYBYTES*2+1, sign->secret_key, n_elements));
+
+  return TRUE;
+
+err:
+#endif /* HAVE_LIBSODIUM */
+  return FALSE;
+}
+
+gboolean ostree_sign_ed25519_set_pk (OstreeSign *self,
+                                     GVariant *public_key,
+                                     GError **error)
+{
+  g_debug ("%s enter", __FUNCTION__);
+  g_return_val_if_fail (OSTREE_IS_SIGN (self), FALSE);
+
+  OstreeSignEd25519 *sign = ostree_sign_ed25519_get_instance_private(OSTREE_SIGN_ED25519(self));
+
+  /* Substitute the key(s) with a new one */
+  if (sign->public_keys != NULL)
+    {
+      g_list_free_full (sign->public_keys, g_object_unref);
+      sign->public_keys = NULL;
+    }
+
+  return ostree_sign_ed25519_add_pk (self, public_key, error);
+}
+
+gboolean ostree_sign_ed25519_add_pk (OstreeSign *self,
+                                     GVariant *public_key,
+                                     GError **error)
+{
+  g_debug ("%s enter", __FUNCTION__);
+  g_return_val_if_fail (OSTREE_IS_SIGN (self), FALSE);
+
+#ifdef HAVE_LIBSODIUM
+  OstreeSignEd25519 *sign = ostree_sign_ed25519_get_instance_private(OSTREE_SIGN_ED25519(self));
+  g_autofree char * hex = NULL;
+  gpointer key = NULL; 
+
+  gsize n_elements = 0;
+  key = (gpointer) g_variant_get_fixed_array (public_key, &n_elements, sizeof(guchar));
+
+  hex = g_malloc0 (crypto_sign_PUBLICKEYBYTES*2 + 1);
+  g_debug ("Read ed25519 public key = %s", sodium_bin2hex (hex, crypto_sign_PUBLICKEYBYTES*2+1, key, n_elements));
+
+  if (n_elements != crypto_sign_PUBLICKEYBYTES)
+    {
+      g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                           "Incorrect ed25519 public key");
+      goto err;
+    }
+
+  key = g_memdup (key, n_elements);
```

Comment:
This construct is weird too :  key from g_variant_get_fixed_array and key from g_memdup, contains the same data, but they don't have same lifetime, so using the same variable to represent both is error prone imho.

And if the key is already in `sign->public_keys`, it is duplicated but not freed. So the `g_memdup` should be inside the if.

Reply:
Nice catch!